source/material_model/rheology/diffusion_dislocation.cc

```diff
@@ -88,99 +89,90 @@ namespace aspect
+              [&](const Vector<double> &evaluation_point,
+                  Vector<double>       &residual)
+            {
+              compute_log_strain_rate_residual(evaluation_point, residual, pressure, temperature, diffusion_creep_parameters, dislocation_creep_parameters, log_edot_ii);
+            };
+
+            nonlinear_solver.setup_jacobian =
+              [&](const Vector<double> &current_u,
+                  const Vector<double> & /*current_f*/)
+            {
+              compute_log_strain_rate_deriv(current_u, pressure, temperature, diffusion_creep_parameters, dislocation_creep_parameters, log_strain_rate_deriv);
+            };
+
+            nonlinear_solver.solve_with_jacobian = [&](const Vector<double> &rhs,
+                                                       Vector<double> &solution,
+                                                       const double /*tolerance*/)
+            {
+              solution(0) -= rhs(0)/log_strain_rate_deriv;
+            };
+
+            nonlinear_solver.solve(log_stress_ii);
+
+            stress_ii = std::exp(log_stress_ii[0]);
             composition_viscosities[j] = std::min(std::max(stress_ii/edot_ii/2, minimum_viscosity), maximum_viscosity);
           }
         return composition_viscosities;
       }
 
+
+      template <int dim>
+      void
+      DiffusionDislocation<dim>::compute_log_strain_rate_deriv(
+        const Vector<double> &evaluation_point,
+        double pressure,
+        double temperature,
+        const Rheology::DiffusionCreepParameters diffusion_creep_parameters,
+        const Rheology::DislocationCreepParameters dislocation_creep_parameters,
+        double &log_strain_rate_deriv) const
+      {
+        const std::pair<double, double> log_diff_edot_and_deriv = diffusion_creep.compute_log_strain_rate_and_derivative(evaluation_point[0], pressure, temperature, diffusion_creep_parameters);
+        const std::pair<double, double> log_disl_edot_and_deriv = dislocation_creep.compute_log_strain_rate_and_derivative(evaluation_point[0], pressure, temperature, dislocation_creep_parameters);
+
+        const double strain_rate_diffusion = std::exp(log_diff_edot_and_deriv.first);
+        const double strain_rate_dislocation = std::exp(log_disl_edot_and_deriv.first);
+        log_strain_rate_deriv = (strain_rate_diffusion * log_diff_edot_and_deriv.second + strain_rate_dislocation * log_disl_edot_and_deriv.second)/
+                                (strain_rate_diffusion + strain_rate_dislocation);
+      }
+
+
+
+      template <int dim>
+      void
+      DiffusionDislocation<dim>::compute_log_strain_rate_residual(
+        const Vector<double> &evaluation_point,
+        Vector<double>       &residual,
+        double pressure,
+        double temperature,
+        const Rheology::DiffusionCreepParameters diffusion_creep_parameters,
+        const Rheology::DislocationCreepParameters dislocation_creep_parameters,
+        double log_edot_ii) const
```

Comment:
Make all input arguments `const` where possible. Same for the previous function.